### PR TITLE
Implement returning arrays from functions

### PR DESCRIPTION
**Pull request by certik**
And use it to implement `numpy.zeros()`. 

TODO:
* [x] The ASR seems to work, but it lowers incorrectly to LLVM it seems, because unlike LFortran, here we assign to a local array first, and then copy to the return array.
* [x] Enable the `test_numpy_02.py` in both LPython and CPython

### CODE REVIEW

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -348,14 +348,20 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                    if (ASR::is_a<ASR::ConstantInteger_t>(*value)) {
+                        ASR::ttype_t *itype = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+                                4, nullptr, 0));
+                        dim.m_start = ASR::down_cast<ASR::expr_t>(ASR::make_ConstantInteger_t(al, loc, 1, itype));
+                        dim.m_end = value;
+                    } else if (ASR::is_a<ASR::Var_t>(*value)) {
+                        ASR::ttype_t *itype = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+                                4, nullptr, 0));
+                        dim.m_start = ASR::down_cast<ASR::expr_t>(ASR::make_ConstantInteger_t(al, loc, 1, itype));
+                        dim.m_end = value;
```
Comment (namannimmo10) [inline src/lpython/semantics/python_ast_to_asr.cpp:360]:
```suggestion
                    if (ASR::is_a<ASR::ConstantInteger_t>(*value) || ASR::is_a<ASR::Var_t>(*value)) {
                        ASR::ttype_t *itype = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
                                4, nullptr, 0));
                        dim.m_start = ASR::down_cast<ASR::expr_t>(ASR::make_ConstantInteger_t(al, loc, 1, itype));
                        dim.m_end = value;
```

Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp:360]: Thanks. It seems this indeed can be done, but I am waiting until I get the LLVM backend working, there is some bug somewhere, as it works in LFortran but not LPython.

**Comment by certik** [comment]
So far the issue is that LFortran's ASR gets transformed into the return array of a function becoming a subroutine's argument, and everything works. In LPython the generated LLVM does not do that and it is indeed incorrect. It's still not clear to me why that happens.

**Comment by certik** [comment]
Found it: it's the `pass_replace_array_op` pass that transforms the function to a subroutine at the ASR level, so the LLVM backend doesn't need to worry about this.

**Comment by certik** [comment]
I think I found the problem:
```diff
--- a/src/libasr/pass/array_op.cpp
+++ b/src/libasr/pass/array_op.cpp
@@ -124,6 +124,8 @@ public:
     void visit_Program(const ASR::Program_t &x) {
         std::vector<std::pair<std::string, ASR::symbol_t*>> replace_vec;
         // Transform nested functions and subroutines
+        // FIXME: this pass only converts local (nested) functions in the program,
+        // but we need to convert all functions
         for (auto &item : x.m_symtab->scope) {
             if (is_a<ASR::Subroutine_t>(*item.second)) {
                 ASR::Subroutine_t *s = down_cast<ASR::Subroutine_t>(item.second);
```

**Comment by certik** [comment]
There is still some bug in the `array_op` logic:
```diff
diff --git a/integration_tests/test_numpy_02.py b/integration_tests/test_numpy_02.py
index a80ba5a27..31e62d581 100644
--- a/integration_tests/test_numpy_02.py
+++ b/integration_tests/test_numpy_02.py
@@ -5,7 +5,7 @@ from numpy import empty
 
 n: i32
 n = TypeVar("n")
-def zeros(n: i32) -> f64[n]:
+def arange(n: i32) -> f64[n]:
     A: f64[n]
     A = empty(n)
     i: i32
@@ -17,7 +17,7 @@ def zeros(n: i32) -> f64[n]:
 
 def test_zeros():
     a: f64[4]
-    a = zeros(4)
+    a = arange(4)
     eps: f64
     eps = 1e-12
     assert abs(a[0] - 0.0) < eps
```
This fails with:
```
$ lpython test_numpy_02.py
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/ondrej/repos/lpython/src/bin/lpython.cpp", line 1544
    err = compile_python_to_object_file(arg_file, tmp_o, runtime_library_dir, compiler_options);
  File "/Users/ondrej/repos/lpython/src/bin/lpython.cpp", line 746
    res = fe.get_llvm3(*asr, diagnostics);
  File "/Users/ondrej/repos/lpython/src/lpython/fortran_evaluator.cpp", line 308
    run_fn);
  File "/Users/ondrej/repos/lpython/src/libasr/codegen/asr_to_llvm.cpp", line 4045
    pass_replace_array_op(al, asr, rl_path);
  File "/Users/ondrej/repos/lpython/src/libasr/pass/array_op.cpp", line 731
    v.visit_TranslationUnit(unit);
  File "/Users/ondrej/repos/lpython/src/libasr/pass/array_op.cpp", line 176
    for (auto &item : x.m_global_scope->scope) {
  File "../libasr/asr.h", line 2279
  File "../libasr/asr.h", line 2124
  File "/Users/ondrej/repos/lpython/src/libasr/pass/array_op.cpp", line 244
    transform_stmts(xx.m_body, xx.n_body, al, array_op_result);
  File "../libasr/pass/pass_utils.h", line 65
  File "../libasr/asr.h", line 2291
  File "../libasr/asr.h", line 2142
  File "/Users/ondrej/repos/lpython/src/libasr/pass/array_op.cpp", line 257
    result_var = x.m_target;
  File "../libasr/asr.h", line 2321
  File "../libasr/asr.h", line 2197
  File "/Users/ondrej/repos/lpython/src/libasr/pass/array_op.cpp", line 421
    head.m_start = PassUtils::get_bound(result_var, i + 1, "lbound", al, unit, rl_path, current_scope);
  File "/Users/ondrej/repos/lpython/src/libasr/pass/pass_utils.cpp", line 272
    unit, rl_path, current_scope, arr_expr->base.loc);
  File "/Users/ondrej/repos/lpython/src/libasr/pass/pass_utils.cpp", line 246
    ASR::Module_t *m = LFortran::ASRUtils::load_module(al, current_scope,
  File "/Users/ondrej/repos/lpython/src/libasr/asr_utils.cpp", line 190
    LFORTRAN_ASSERT(asr_verify(*tu));
  File "/Users/ondrej/repos/lpython/src/libasr/asr_verify.cpp", line 428
    v.visit_TranslationUnit(unit);
  File "/Users/ondrej/repos/lpython/src/libasr/asr_verify.cpp", line 110
    for (auto &a : x.m_global_scope->scope) {
  File "../libasr/asr.h", line 2279
  File "../libasr/asr.h", line 2124
  File "/Users/ondrej/repos/lpython/src/libasr/asr_verify.cpp", line 208
    visit_stmt(*x.m_body[i]);
  File "../libasr/asr.h", line 2291
  File "../libasr/asr.h", line 2142
  File "../libasr/asr.h", line 2463
  File "../libasr/asr.h", line 2321
  File "../libasr/asr.h", line 2184
  File "/Users/ondrej/repos/lpython/src/libasr/asr_verify.cpp", line 383
    x.base.base.loc);
  File "/Users/ondrej/repos/lpython/src/libasr/asr_verify.cpp", line 60
    require(cond, msg);
  File "/Users/ondrej/repos/lpython/src/libasr/asr_verify.cpp", line 49
    throw LFortranException("ASR verify failed: " + error_msg);
LFortranException: ASR verify failed: 345:353: FunctionCall::m_name `arange` cannot point outside of its symbol table
```
That's due to the fact that the change of name changes the order the functions are visited, and something just doesn't quite work yet. We need to fix that.

File: src/libasr/asr_utils.cpp [hunk 2]
```diff
@@ -187,7 +187,10 @@ ASR::Module_t* load_module(Allocator &al, SymbolTable *symtab,
 
     // Fix all external symbols
     fix_external_symbols(*tu, *symtab);
-    LFORTRAN_ASSERT(asr_verify(*tu));
+    // FIXME: this verify must be turned of while running from inside
+    // `array_op`, until the pass is complete, then it will pass again,
+    // otherwise it fails, because during the pass the ASR is not valid
+    //LFORTRAN_ASSERT(asr_verify(*tu));
```
Comment (certik) [inline src/libasr/asr_utils.cpp:193]: We need to figure out how to fix this.

Comment (certik) [reply src/libasr/asr_utils.cpp]: Fixed in d700b850f2ae20713a5896b49e0def20462e734a.

**Comment by certik** [comment]
I think this is finally good to go in!

**Pull request merged**